### PR TITLE
Allow empty string as valid setting for launch_type in CloudWatchEventTarget

### LINKED ISSUE #16078: resource/aws_cloudwatch_event_target: ecs_target's launch_type should not have a default
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version

```
Terraform v0.13.5
+ provider registry.terraform.io/hashicorp/aws v3.14.1
```

### Affected Resource(s)

* aws_cloudwatch_event_target

### Terraform Configuration Files

```hcl
resource "aws_ecs_cluster" "database" {
  name = "database"
}

resource "aws_cloudwatch_event_rule" "db_backup" {
  name                = "DBBackup"
  description         = "Run daily at 07:05 UTC"
  schedule_expression = "cron(5 7 * * ? *)"
}

resource "aws_cloudwatch_event_target" "dump_database" {
  arn       = aws_ecs_cluster.database.arn
  rule      = aws_cloudwatch_event_rule.db_backup.name
  role_arn  = "arn:aws:iam::696278446003:role/ecsEventsRole"

  ecs_target {
    task_count          = 1
    task_definition_arn = aws_ecs_task_definition.dump_database.arn
  }
}

resource "aws_ecs_task_definition" "dump_database" {
  family                = "service"
  container_definitions = file("task-definitions/dump_database.json")
}
```

task-definitions/dump_database.json:
```json
[
  {
    "name": "first",
    "image": "service-first",
    "cpu": 10,
    "memory": 512,
    "essential": true,
    "portMappings": [
      {
        "containerPort": 80,
        "hostPort": 80
      }
    ]
  }
]
```
### Expected Behavior

The `launch_type` should not be set.

### Actual Behavior

In the `aws_cloudwatch_event_target` output,  the `ecs_target` `launch_type` is set to `EC2`.

### Steps to Reproduce

1. `terraform apply`

### References

* [AWS CloudWatch Events ECS Parameters](https://docs.aws.amazon.com/eventbridge/latest/APIReference/API_EcsParameters.html)

### PR DESCRIPTION
**Pull request by ianneub**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates #16078

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSCloudWatchEventTarget_'

==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSCloudWatchEventTarget_ -timeout 180m
=== RUN   TestAccAWSCloudWatchEventTarget_basic
=== PAUSE TestAccAWSCloudWatchEventTarget_basic
=== RUN   TestAccAWSCloudWatchEventTarget_EventBusName
=== PAUSE TestAccAWSCloudWatchEventTarget_EventBusName
=== RUN   TestAccAWSCloudWatchEventTarget_GeneratedTargetId
=== PAUSE TestAccAWSCloudWatchEventTarget_GeneratedTargetId
=== RUN   TestAccAWSCloudWatchEventTarget_RetryPolicy_DeadLetterConfig
=== PAUSE TestAccAWSCloudWatchEventTarget_RetryPolicy_DeadLetterConfig
=== RUN   TestAccAWSCloudWatchEventTarget_full
=== PAUSE TestAccAWSCloudWatchEventTarget_full
=== RUN   TestAccAWSCloudWatchEventTarget_disappears
=== PAUSE TestAccAWSCloudWatchEventTarget_disappears
=== RUN   TestAccAWSCloudWatchEventTarget_ssmDocument
=== PAUSE TestAccAWSCloudWatchEventTarget_ssmDocument
=== RUN   TestAccAWSCloudWatchEventTarget_http
=== PAUSE TestAccAWSCloudWatchEventTarget_http
=== RUN   TestAccAWSCloudWatchEventTarget_ecs
=== PAUSE TestAccAWSCloudWatchEventTarget_ecs
=== RUN   TestAccAWSCloudWatchEventTarget_ecsWithBlankTaskCount
=== PAUSE TestAccAWSCloudWatchEventTarget_ecsWithBlankTaskCount
=== RUN   TestAccAWSCloudWatchEventTarget_batch
=== PAUSE TestAccAWSCloudWatchEventTarget_batch
=== RUN   TestAccAWSCloudWatchEventTarget_kinesis
=== PAUSE TestAccAWSCloudWatchEventTarget_kinesis
=== RUN   TestAccAWSCloudWatchEventTarget_sqs
=== PAUSE TestAccAWSCloudWatchEventTarget_sqs
=== RUN   TestAccAWSCloudWatchEventTarget_input_transformer
=== PAUSE TestAccAWSCloudWatchEventTarget_input_transformer
=== RUN   TestAccAWSCloudWatchEventTarget_inputTransformerJsonString
=== PAUSE TestAccAWSCloudWatchEventTarget_inputTransformerJsonString
=== RUN   TestAccAWSCloudWatchEventTarget_PartnerEventBus
    resource_aws_cloudwatch_event_target_test.go:673: Environment variable EVENT_BRIDGE_PARTNER_EVENT_BUS_NAME is not set
--- SKIP: TestAccAWSCloudWatchEventTarget_PartnerEventBus (0.00s)
=== CONT  TestAccAWSCloudWatchEventTarget_basic
=== CONT  TestAccAWSCloudWatchEventTarget_ecs
=== CONT  TestAccAWSCloudWatchEventTarget_kinesis
=== CONT  TestAccAWSCloudWatchEventTarget_full
=== CONT  TestAccAWSCloudWatchEventTarget_http
=== CONT  TestAccAWSCloudWatchEventTarget_ssmDocument
=== CONT  TestAccAWSCloudWatchEventTarget_disappears
=== CONT  TestAccAWSCloudWatchEventTarget_sqs
=== CONT  TestAccAWSCloudWatchEventTarget_inputTransformerJsonString
=== CONT  TestAccAWSCloudWatchEventTarget_input_transformer
=== CONT  TestAccAWSCloudWatchEventTarget_RetryPolicy_DeadLetterConfig
=== CONT  TestAccAWSCloudWatchEventTarget_EventBusName
=== CONT  TestAccAWSCloudWatchEventTarget_GeneratedTargetId
=== CONT  TestAccAWSCloudWatchEventTarget_batch
=== CONT  TestAccAWSCloudWatchEventTarget_ecsWithBlankTaskCount
--- PASS: TestAccAWSCloudWatchEventTarget_GeneratedTargetId (59.81s)
--- PASS: TestAccAWSCloudWatchEventTarget_http (93.22s)
--- PASS: TestAccAWSCloudWatchEventTarget_sqs (141.61s)
--- PASS: TestAccAWSCloudWatchEventTarget_ecsWithBlankTaskCount (156.11s)
--- PASS: TestAccAWSCloudWatchEventTarget_input_transformer (160.64s)
--- PASS: TestAccAWSCloudWatchEventTarget_disappears (162.59s)
--- PASS: TestAccAWSCloudWatchEventTarget_RetryPolicy_DeadLetterConfig (163.23s)
--- PASS: TestAccAWSCloudWatchEventTarget_full (163.35s)
--- PASS: TestAccAWSCloudWatchEventTarget_inputTransformerJsonString (166.32s)
--- PASS: TestAccAWSCloudWatchEventTarget_ecs (167.78s)
--- PASS: TestAccAWSCloudWatchEventTarget_EventBusName (168.02s)
--- PASS: TestAccAWSCloudWatchEventTarget_ssmDocument (174.26s)
--- PASS: TestAccAWSCloudWatchEventTarget_kinesis (177.12s)
--- PASS: TestAccAWSCloudWatchEventTarget_basic (186.24s)
--- PASS: TestAccAWSCloudWatchEventTarget_batch (248.19s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	251.042s
```

### CODE REVIEW

File: aws/resource_aws_cloudwatch_event_target.go [hunk 1]
```diff
@@ -146,7 +146,7 @@ func resourceAwsCloudWatchEventTarget() *schema.Resource {
 							Type:         schema.TypeString,
 							Optional:     true,
 							Default:      events.LaunchTypeEc2,
-							ValidateFunc: validation.StringInSlice(events.LaunchType_Values(), false),
+							ValidateFunc: validation.StringInSlice(append(events.LaunchType_Values(), []string{""}...), false),
```
Comment (anGie44) [inline aws/resource_aws_cloudwatch_event_target.go:149]: Exactly the right place, but I would just consolidate this by using our `validation.Any` helper method 
```suggestion
							ValidateFunc: validation.Any(
								validation.StringIsEmpty,
								validation.StringInSlice(events.LaunchType_Values(), false),
							),
```

File: .changelog/19555.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+```release-note:bug
+resource/aws_cloudwatch_event_target: Fix ecs_target.launch_type not allowing empty string values.
```
Comment (anGie44) [inline .changelog/19555.txt:2]: Nit: backticks often help for visibility
```suggestion
resource/aws_cloudwatch_event_target: Fix `ecs_target.launch_type` not allowing empty string values.
```

**Review by anGie44 — CHANGES_REQUESTED** [review]
Hi @ianneub, this is off to a great start! Just small comments to address. 

Also, do you mind adding in a new acceptance test to validate creation with an empty string as well as updating from non-empty to empty? Feel free to reach out if you have any questions or if you're not readily available at this time, just drop a note and I'd be happy to continue the work here 😃

File: aws/resource_aws_cloudwatch_event_target_test.go [hunk 3]
```diff
@@ -447,6 +447,49 @@ func TestAccAWSCloudWatchEventTarget_ecs(t *testing.T) {
 				ImportStateIdFunc: testAccAWSCloudWatchEventTargetImportStateIdFunc(resourceName),
 				ImportStateVerify: true,
 			},
+			{
+				Config: testAccAWSCloudWatchEventTargetConfigEcsWithBlankLaunchType(rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckCloudWatchEventTargetExists(resourceName, &v),
+					resource.TestCheckResourceAttr(resourceName, "ecs_target.0.launch_type", ""),
+				),
+			},
```
Comment (ianneub) [inline aws/resource_aws_cloudwatch_event_target_test.go:456]: @anGie44 Is this an acceptable place for this test to ensure changing a set LaunchType to blank will work?

**Comment by ianneub** [comment]
Thanks for checking this PR out @anGie44 . I've updated the PR with the following changes:

1. A new acceptance test and a tweak to the existing test
2. Tweaked the changelog entry
3. I had to modify your change to `aws/resource_aws_cloudwatch_event_target.go` on lines 146 - 148 as gofmt's was complaining about the indentation.

Please let me know if you have any suggestions to improve the PR.

After running my tests I get the following output:

```
$ ACCTEST_PARALLELISM=1 AWS_PROFILE=default make testacc TESTARGS='-run=TestAccAWSCloudWatchEventTarget_ecs'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 1 -run=TestAccAWSCloudWatchEventTarget_ecs -timeout 180m
=== RUN   TestAccAWSCloudWatchEventTarget_ecs
=== PAUSE TestAccAWSCloudWatchEventTarget_ecs
=== RUN   TestAccAWSCloudWatchEventTarget_ecsWithBlankLaunchType
=== PAUSE TestAccAWSCloudWatchEventTarget_ecsWithBlankLaunchType
=== RUN   TestAccAWSCloudWatchEventTarget_ecsWithBlankTaskCount
=== PAUSE TestAccAWSCloudWatchEventTarget_ecsWithBlankTaskCount
=== CONT  TestAccAWSCloudWatchEventTarget_ecs
--- PASS: TestAccAWSCloudWatchEventTarget_ecs (37.67s)
=== CONT  TestAccAWSCloudWatchEventTarget_ecsWithBlankTaskCount
--- PASS: TestAccAWSCloudWatchEventTarget_ecsWithBlankTaskCount (25.03s)
=== CONT  TestAccAWSCloudWatchEventTarget_ecsWithBlankLaunchType
--- PASS: TestAccAWSCloudWatchEventTarget_ecsWithBlankLaunchType (24.97s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	89.522s
```

**Comment by ianneub** [comment]
Thanks @anGie44 . I believe those changes are good to go, however this is my first commit to Terraform. Please let me know if you have any other suggestions to make it better.

**Review by anGie44 — APPROVED** [review]
Great work @ianneub ! Only non-code change i'll make is update the description to `Relate` to the original issue instead of `Fix` since i think the issue is a valid breaking-change we should address in a major release.

Output of acceptance tests (commercial):
```
--- SKIP: TestAccAWSCloudWatchEventTarget_PartnerEventBus (0.00s)
--- PASS: TestAccAWSCloudWatchEventTarget_disappears (35.61s)
--- PASS: TestAccAWSCloudWatchEventTarget_sqs (39.84s)
--- PASS: TestAccAWSCloudWatchEventTarget_GeneratedTargetId (42.10s)
--- PASS: TestAccAWSCloudWatchEventTarget_ssmDocument (45.21s)
--- PASS: TestAccAWSCloudWatchEventTarget_http (48.44s)
--- PASS: TestAccAWSCloudWatchEventTarget_inputTransformerJsonString (56.68s)
--- PASS: TestAccAWSCloudWatchEventTarget_ecs (58.01s)
--- PASS: TestAccAWSCloudWatchEventTarget_ecsWithBlankTaskCount (58.04s)
--- PASS: TestAccAWSCloudWatchEventTarget_EventBusName (62.05s)
--- PASS: TestAccAWSCloudWatchEventTarget_input_transformer (68.45s)
--- PASS: TestAccAWSCloudWatchEventTarget_basic (73.92s)
--- PASS: TestAccAWSCloudWatchEventTarget_RetryPolicy_DeadLetterConfig (78.43s)
--- PASS: TestAccAWSCloudWatchEventTarget_full (80.06s)
--- PASS: TestAccAWSCloudWatchEventTarget_kinesis (81.19s)
--- PASS: TestAccAWSCloudWatchEventTarget_ecsWithBlankLaunchType (89.42s)
--- PASS: TestAccAWSCloudWatchEventTarget_batch (100.35s)
```

**Update:** I missed to mention we should update the documentation as well to be more explicit about this empty string option! I'll add this in 👍
